core/src/main/java/org/infinispan/interceptors/TxInterceptor.java

```diff
@@ -235,18 +249,73 @@ public Object visitSizeCommand(InvocationContext ctx, SizeCommand command) throw
    }
 
    @Override
-   public Object visitKeySetCommand(InvocationContext ctx, KeySetCommand command) throws Throwable {
-      return enlistReadAndInvokeNext(ctx, command);
-   }
-
-   @Override
-   public Object visitValuesCommand(InvocationContext ctx, ValuesCommand command) throws Throwable {
-      return enlistReadAndInvokeNext(ctx, command);
+   public CacheSet<K> visitKeySetCommand(InvocationContext ctx, KeySetCommand command) throws Throwable {
+      CacheSet<K> set = (CacheSet<K>) enlistReadAndInvokeNext(ctx, command);
+      if (ctx.isInTxScope()) {
```

Comment:
Hmmm, do we really want to support keeping a transaction running while the user is iterating over the entire cache? Tx awareness seems ok in operations like `cache.keySet().contains(k)`, but for bulk operations it seems like nothing but trouble.